File: sdk/php/tests/ConformanceTest.php
<?php

declare(strict_types=1);

namespace Cordum\Cap\Tests;

use Cordum\Agent\V1\BusPacket;
use Cordum\Agent\V1\JobPriority;
use Cordum\Agent\V1\JobStatus;
use Cordum\Agent\V1\AlertSeverity;
use Cordum\Agent\V1\ComponentRole;
use Cordum\Cap\Codec;
use Cordum\Cap\Signing;
use OpenSSLAsymmetricKey;
use PHPUnit\Framework\TestCase;

class ConformanceTest extends TestCase
{
    private static function fixturesDir(): string
    {
        return dirname(__DIR__, 3) . '/spec/conformance/fixtures';
    }

    private static function loadFixture(string $name): BusPacket
    {
        $path = self::fixturesDir() . '/' . $name;
        $data = file_get_contents($path);
        if ($data === false) {
            throw new \RuntimeException("Failed to read fixture: {$name}");
        }
        $pkt = new BusPacket();
        $pkt->mergeFromString($data);
        return $pkt;
    }

    private static function loadConformancePublicKey(): OpenSSLAsymmetricKey
    {
        $path = self::fixturesDir() . '/public_key.pem';
        $pem = file_get_contents($path);
        return Signing::loadPublicKey($pem);
    }

    private function verifyFixtureSignature(BusPacket $pkt, OpenSSLAsymmetricKey $key): void
    {
        $this->assertNotEmpty($pkt->getSignature(), 'fixture packet has no signature');
        $this->assertTrue(Signing::verifyPacketSignature($pkt, $key), 'signature verification failed');
    }

    private function assertTimestamp(BusPacket $pkt): void
    {
        $ts = $pkt->getCreatedAt();
        $this->assertNotNull($ts);
        $this->assertSame(1704164645, $ts->getSeconds());
        $this->assertSame(0, $ts->getNanos());
        $this->assertSame(1, $pkt->getProtocolVersion());
    }

    #[\PHPUnit\Framework\Attributes\Test]
    public function conformanceJobRequest(): void
    {
        $key = self::loadConformancePublicKey();
        $pkt = self::loadFixture('buspacket_job_request.bin');

        $this->verifyFixtureSignature($pkt, $key);
        $this->assertTimestamp($pkt);

        $req = $pkt->getJobRequest();
        $this->assertNotNull($req);
        $this->assertSame('job-req-1', $req->getJobId());
        $this->assertSame('job.tools', $req->getTopic());
        $this->assertSame(JobPriority::JOB_PRIORITY_INTERACTIVE, $req->getPriority());
        $this->assertSame('redis://ctx:job-req-1', $req->getContextPtr());
    }

    #[\PHPUnit\Framework\Attributes\Test]
    public function conformanceJobResult(): void
    {
        $key = self::loadConformancePublicKey();
        $pkt = self::loadFixture('buspacket_job_result.bin');

        $this->verifyFixtureSignature($pkt, $key);
        $this->assertTimestamp($pkt);

        $res = $pkt->getJobResult();
        $this->assertNotNull($res);
        $this->assertSame('job-res-1', $res->getJobId());
        $this->assertSame('worker-1', $res->getWorkerId());
        $this->assertSame(JobStatus::JOB_STATUS_FAILED_RETRYABLE, $res->getStatus());
        $this->assertSame('E_TEMP', $res->getErrorCode());
        $this->assertCount(2, $res->getArtifactPtrs());
    }

    #[\PHPUnit\Framework\Attributes\Test]
    public function conformanceHeartbeat(): void
    {
        $key = self::loadConformancePublicKey();
        $pkt = self::loadFixture('buspacket_heartbeat.bin');

        $this->verifyFixtureSignature($pkt, $key);
        $this->assertTimestamp($pkt);

        $hb = $pkt->getHeartbeat();
        $this->assertNotNull($hb);
        $this->assertSame('worker-1', $hb->getWorkerId());
        $this->assertSame('job.tools', $hb->getPool());
        $this->assertSame('us-east-1a', $hb->getLabels()['zone']);
    }

    #[\PHPUnit\Framework\Attributes\Test]
    public function conformanceJobProgress(): void
    {
        $key = self::loadConformancePublicKey();
        $pkt = self::loadFixture('buspacket_job_progress.bin');

        $this->verifyFixtureSignature($pkt, $key);
        $this->assertTimestamp($pkt);

        $progress = $pkt->getJobProgress();
        $this->assertNotNull($progress);
        $this->assertSame('job-prog-1', $progress->getJobId());
        $this->assertSame(50, $progress->getPercent());
        $this->assertSame(JobStatus::JOB_STATUS_RUNNING, $progress->getStatus());
    }

    #[\PHPUnit\Framework\Attributes\Test]
    public function conformanceJobCancel(): void
    {
        $key = self::loadConformancePublicKey();
        $pkt = self::loadFixture('buspacket_job_cancel.bin');

        $this->verifyFixtureSignature($pkt, $key);
        $this->assertTimestamp($pkt);

        $cancel = $pkt->getJobCancel();
        $this->assertNotNull($cancel);
        $this->assertSame('job-cancel-1', $cancel->getJobId());
        $this->assertSame('user-7', $cancel->getRequestedBy());
    }

    #[\PHPUnit\Framework\Attributes\Test]
    public function conformanceAlert(): void
    {
        $key = self::loadConformancePublicKey();
        $pkt = self::loadFixture('buspacket_alert.bin');

        $this->verifyFixtureSignature($pkt, $key);
        $this->assertTimestamp($pkt);

        $alert = $pkt->getAlert();
        $this->assertNotNull($alert);
        $this->assertSame('WARN', $alert->getLevel());
        $this->assertSame('scheduler', $alert->getComponent());
    }

    #[\PHPUnit\Framework\Attributes\Test]
    public function conformanceHandshake(): void
    {
        $key = self::loadConformancePublicKey();
        $pkt = self::loadFixture('buspacket_handshake.bin');

        $this->verifyFixtureSignature($pkt, $key);
        $this->assertTimestamp($pkt);

        $hs = $pkt->getHandshake();
        $this->assertNotNull($hs);
        $this->assertSame('worker-1', $hs->getComponentId());
        $this->assertSame(ComponentRole::COMPONENT_ROLE_WORKER, $hs->getRole());
        $this->assertSame('2.0.19', $hs->getSdkVersion());
    }

    #[\PHPUnit\Framework\Attributes\Test]
    public function conformanceAlertEnhanced(): void
    {
        $key = self::loadConformancePublicKey();
        $pkt = self::loadFixture('buspacket_alert_enhanced.bin');

        $this->verifyFixtureSignature($pkt, $key);
        $this->assertTimestamp($pkt);

        $alert = $pkt->getAlert();
        $this->assertNotNull($alert);
        $this->assertSame('CRITICAL', $alert->getLevel());
        $this->assertSame('scheduler', $alert->getComponent());
        $this->assertSame('SIGNATURE_INVALID', $alert->getCode());
        $this->assertSame(AlertSeverity::ALERT_SEVERITY_CRITICAL, $alert->getSeverity());
        $this->assertSame('scheduler-1', $alert->getSourceComponent());
        $this->assertSame('worker-bad', $alert->getDetails()['sender']);
        $this->assertSame('sys.job.result', $alert->getDetails()['subject']);
        $this->assertSame('trace-offending-packet', $alert->getTraceId());
    }

    #[\PHPUnit\Framework\Attributes\Test]
    public function conformanceAllFixturesSignatureVerification(): void
    {
        $key = self::loadConformancePublicKey();
        $fixtures = [
            'buspacket_job_request.bin',
            'buspacket_job_result.bin',
            'buspacket_heartbeat.bin',
            'buspacket_job_progress.bin',
            'buspacket_job_cancel.bin',
            'buspacket_alert.bin',
            'buspacket_handshake.bin',
            'buspacket_alert_enhanced.bin',
        ];

        foreach ($fixtures as $fixture) {
            $pkt = self::loadFixture($fixture);
            $this->assertNotEmpty($pkt->getSignature(), "{$fixture} has no signature");

            $unsignedBytes = Codec::marshalUnsignedForSignature($pkt);
            $verified = openssl_verify($unsignedBytes, $pkt->getSignature(), $key, OPENSSL_ALGO_SHA256) === 1;
            $this->assertTrue($verified, "{$fixture} signature verification failed");
        }
    }
}

File: sdk/rust/tests/conformance_test.rs
use std::fs;
use std::path::{Path, PathBuf};

use cap_sdk::codec::marshal_unsigned_for_signature;
use cap_sdk::pb::{bus_packet::Payload, BusPacket};
use cap_sdk::signing::load_public_key;
use p256::ecdsa::VerifyingKey;
use prost::Message;
use ecdsa::signature::Verifier;

fn fixtures_dir() -> PathBuf {
    let manifest_dir = Path::new(env!("CARGO_MANIFEST_DIR"));
    manifest_dir
        .parent()
        .unwrap()
        .parent()
        .unwrap()
        .join("spec")
        .join("conformance")
        .join("fixtures")
}

fn load_fixture(name: &str) -> BusPacket {
    let path = fixtures_dir().join(name);
    let data = fs::read(&path).unwrap_or_else(|e| panic!("read fixture {}: {}", name, e));
    BusPacket::decode(data.as_slice())
        .unwrap_or_else(|e| panic!("decode fixture {}: {}", name, e))
}

fn load_conformance_public_key() -> VerifyingKey {
    let path = fixtures_dir().join("public_key.pem");
    let pem = fs::read_to_string(&path).expect("read public_key.pem");
    load_public_key(&pem).expect("parse public key")
}

fn verify_fixture_signature(pkt: &BusPacket, key: &VerifyingKey) {
    assert!(
        !pkt.signature.is_empty(),
        "fixture packet has no signature"
    );
    cap_sdk::signing::verify_packet_signature(pkt, key)
        .expect("signature verification failed");
}

fn assert_timestamp(pkt: &BusPacket) {
    let ts = pkt.created_at.as_ref().expect("missing created_at");
    // Expected: 2024-01-02T03:04:05Z
    assert_eq!(ts.seconds, 1704164645, "unexpected created_at seconds");
    assert_eq!(ts.nanos, 0, "unexpected created_at nanos");
    assert_eq!(pkt.protocol_version, 1, "unexpected protocol_version");
}

#[test]
fn conformance_job_request() {
    let key = load_conformance_public_key();
    let pkt = load_fixture("buspacket_job_request.bin");

    verify_fixture_signature(&pkt, &key);
    assert_timestamp(&pkt);

    let req = match &pkt.payload {
        Some(Payload::JobRequest(r)) => r,
        _ => panic!("expected JobRequest payload"),
    };

    assert_eq!(req.job_id, "job-req-1");
    assert_eq!(req.topic, "job.tools");
    assert_eq!(req.context_ptr, "redis://ctx:job-req-1");
}

#[test]
fn conformance_job_result() {
    let key = load_conformance_public_key();
    let pkt = load_fixture("buspacket_job_result.bin");

    verify_fixture_signature(&pkt, &key);
    assert_timestamp(&pkt);

    let res = match &pkt.payload {
        Some(Payload::JobResult(r)) => r,
        _ => panic!("expected JobResult payload"),
    };

    assert_eq!(res.job_id, "job-res-1");
    assert_eq!(res.worker_id, "worker-1");
    assert_eq!(res.error_code, "E_TEMP");
    assert_eq!(res.artifact_ptrs.len(), 2);
}

#[test]
fn conformance_heartbeat() {
    let key = load_conformance_public_key();
    let pkt = load_fixture("buspacket_heartbeat.bin");

    verify_fixture_signature(&pkt, &key);
    assert_timestamp(&pkt);

    let hb = match &pkt.payload {
        Some(Payload::Heartbeat(h)) => h,
        _ => panic!("expected Heartbeat payload"),
    };

    assert_eq!(hb.worker_id, "worker-1");
    assert_eq!(hb.pool, "job.tools");
    assert_eq!(hb.labels.get("zone").map(String::as_str), Some("us-east-1a"));
}

#[test]
fn conformance_job_progress() {
    let key = load_conformance_public_key();
    let pkt = load_fixture("buspacket_job_progress.bin");

    verify_fixture_signature(&pkt, &key);
    assert_timestamp(&pkt);

    let progress = match &pkt.payload {
        Some(Payload::JobProgress(p)) => p,
        _ => panic!("expected JobProgress payload"),
    };

    assert_eq!(progress.job_id, "job-prog-1");
    assert_eq!(progress.percent, 50);
}

#[test]
fn conformance_job_cancel() {
    let key = load_conformance_public_key();
    let pkt = load_fixture("buspacket_job_cancel.bin");

    verify_fixture_signature(&pkt, &key);
    assert_timestamp(&pkt);

    let cancel = match &pkt.payload {
        Some(Payload::JobCancel(c)) => c,
        _ => panic!("expected JobCancel payload"),
    };

    assert_eq!(cancel.job_id, "job-cancel-1");
    assert_eq!(cancel.requested_by, "user-7");
}

#[test]
fn conformance_alert() {
    let key = load_conformance_public_key();
    let pkt = load_fixture("buspacket_alert.bin");

    verify_fixture_signature(&pkt, &key);
    assert_timestamp(&pkt);

    let alert = match &pkt.payload {
        Some(Payload::Alert(a)) => a,
        _ => panic!("expected Alert payload"),
    };

    assert_eq!(alert.level, "WARN");
    assert_eq!(alert.component, "scheduler");
}

#[test]
fn conformance_handshake() {
    let key = load_conformance_public_key();
    let pkt = load_fixture("buspacket_handshake.bin");

    verify_fixture_signature(&pkt, &key);
    assert_timestamp(&pkt);

    let hs = match &pkt.payload {
        Some(Payload::Handshake(h)) => h,
        _ => panic!("expected Handshake payload"),
    };

    assert_eq!(hs.component_id, "worker-1");
    assert_eq!(hs.supported_versions, vec![1]);
    assert_eq!(hs.sdk_version, "2.0.19");
    assert_eq!(hs.capabilities.get("signatures"), Some(&true));
    assert_eq!(hs.capabilities.get("progress"), Some(&true));
    assert_eq!(hs.capabilities.get("cancel"), Some(&true));
}

#[test]
fn conformance_alert_enhanced() {
    let key = load_conformance_public_key();
    let pkt = load_fixture("buspacket_alert_enhanced.bin");

    verify_fixture_signature(&pkt, &key);
    assert_timestamp(&pkt);

    let alert = match &pkt.payload {
        Some(Payload::Alert(a)) => a,
        _ => panic!("expected Alert payload"),
    };

    assert_eq!(alert.level, "CRITICAL");
    assert_eq!(alert.component, "scheduler");
    assert_eq!(alert.code, "SIGNATURE_INVALID");
    assert_eq!(alert.source_component, "scheduler-1");
    assert_eq!(
        alert.details.get("sender").map(String::as_str),
        Some("worker-bad")
    );
    assert_eq!(
        alert.details.get("subject").map(String::as_str),
        Some("sys.job.result")
    );
    assert_eq!(alert.trace_id, "trace-offending-packet");
}

#[test]
fn conformance_unsigned_bytes_match_across_fixtures() {
    let key = load_conformance_public_key();

    // For each fixture, verify that stripping signature and re-serializing
    // produces bytes whose SHA-256 hash can be verified against the signature
    let fixtures = [
        "buspacket_job_request.bin",
        "buspacket_job_result.bin",
        "buspacket_heartbeat.bin",
        "buspacket_job_progress.bin",
        "buspacket_job_cancel.bin",
        "buspacket_alert.bin",
        "buspacket_handshake.bin",
        "buspacket_alert_enhanced.bin",
    ];

    for fixture_name in &fixtures {
        let pkt = load_fixture(fixture_name);
        assert!(
            !pkt.signature.is_empty(),
            "{} has no signature",
            fixture_name
        );

        // Verify the full roundtrip: unsigned bytes -> verify with public key
        // VerifyingKey::verify() internally hashes with SHA-256 before verifying
        let unsigned_bytes = marshal_unsigned_for_signature(&pkt);

        let sig = p256::ecdsa::DerSignature::try_from(pkt.signature.as_slice())
            .unwrap_or_else(|e| panic!("{} DER parse: {}", fixture_name, e));

        key.verify(&unsigned_bytes, &sig)
            .unwrap_or_else(|e| panic!("{} verify: {}", fixture_name, e));
    }
}
